dev/Kozani/KozaniRemoteManager/KozaniDvcServer.cpp

```diff
@@ -0,0 +1,734 @@
+        }
+
+        m_errorReportingEnabled = true;
+        return S_OK;
+    }
+
+    void KozaniDvcServer::OpenDvc()
+    {
+        auto lock{ std::unique_lock<std::recursive_mutex>(m_dvcLock) };
+
+        using unique_channel_handle = wil::unique_any_handle_null<decltype(&::WTSVirtualChannelClose), ::WTSVirtualChannelClose>;
+        using unique_channel_memory = wil::unique_any<void*, decltype(&::WTSFreeMemory), ::WTSFreeMemory>;
+
+        unique_channel_handle wtsHandle(::WTSVirtualChannelOpenEx(
+            WTS_CURRENT_SESSION,
+            const_cast<LPSTR>(DvcChannelName),
+            WTS_CHANNEL_OPTION_DYNAMIC));
+
+        if (wtsHandle == nullptr)
+        {
+            // DVC channel didn't exist - the current session is not a remote session or the client of this remote session has not
+            // created a listener of this DVC channel. It is required that the client has created a DVC channel listener to listen
+            // to this specific channel name before the server can open the named DVC channle. Otherwise, ERROR_GEN_FAILURE (31) 
+            // may be returned.
+            LogDebugMessage("WTSVirtualChannelOpenEx failed with error: %u", GetLastError());
+            winrt::throw_last_error();
+        }
+
+        unique_channel_memory fileHandle;
+        DWORD len{};
+
+        THROW_LAST_ERROR_IF(!::WTSVirtualChannelQuery(
+            wtsHandle.get(),
+            WTSVirtualFileHandle,
+            fileHandle.put(),
+            &len));
+        THROW_HR_IF(E_UNEXPECTED, (len != sizeof(HANDLE)));
+
+        THROW_IF_WIN32_BOOL_FALSE(::DuplicateHandle(
+            ::GetCurrentProcess(),
+            *reinterpret_cast<HANDLE*>(fileHandle.get()),
+            ::GetCurrentProcess(),
+            m_dvcHandle.put(),
+            0,
+            FALSE,
+            DUPLICATE_SAME_ACCESS));
+    }
+
+    // This function will be executed in a separate thread from the main thread and shouldn't throw any exceptions.
+    // Any uncaught exceptions will cause the process to terminate.
+    void KozaniDvcServer::ProcessDvcIncomingDataThreadFunc() noexcept try
+    {
+        HRESULT hr{ LOG_IF_FAILED(ProcessDvcIncomingData()) };
+        if (FAILED(hr))
+        {
+            DvcListenerThreadExiting(hr, "Failed to process incoming DVC data");
+        }
+    }
+    CATCH_LOG_RETURN()
+
+    HRESULT KozaniDvcServer::ProcessDvcIncomingData() noexcept try
+    {
+        bool exitThread = false;
+        wil::unique_event_nothrow readComplete;
+        HRESULT hr = readComplete.create();
+        if (FAILED(hr))
+        {
+            DvcListenerThreadExiting(hr, "Failed to create readComplete event");
+            RETURN_HR(hr);
+        }
+
+        // CHANNEL_PDU_LENGTH is the max amount of data sent/received in one DVC operation. Data larger than that is segmented into chunks of 
+        // that size and sent/received as multiple operations.
+        BYTE readBuffer[CHANNEL_PDU_LENGTH];
```

Comment:
Intentionally not to clear a big buffer, which is more than 1600 bytes.

Reply:
That is a security hole if ever accessed before initialization, though. Given the recent push for rust, I think we should strive for memory safety over little performance here if the buffer is only ever initialized once in a while.

Reply:
>Intentionally not to clear a big buffer

Sorry @rhuang-msft but that perf is insignificant. Not initializing locals is a violation of the coding guidelines, and for good reasons.

Either it's too big for a stack variable so dynamically allocate memory, or it's small enough for a stack variable and always initialize it. Here's the relevant coding guideline:


***OUT.20: Always immediately initialize or zero output parameters***

**Reason**

Security, repeatability and convention. Uninitialized data leads to security violations and unpredictable, rare and hard to reproduce bugs.

**Notes**

Callers of output parameter returning functions should always initialize variables passed to output parameters

Reply:
There are exceptions for arrays, otherwise one will not be able to write efficient code, or worse, always write in-efficient code. I understand the risk of uninitialized data but in this case, it is very simple usage and obviously will not cause any uninitialized data access issues. Initializing large array unnecessarily is bad practice. Say, if you have a 10MB buffer waiting to be filled, do you waste the resources to first fill it with 0s? In general, there are trade-offs and each case should be evaluated and see which pattern should be used. In this case, it is a simple out-buffer to be filled immediately by the ReadFile call and the buffer will be allocated and release for each ProcessDvcIncomingData() call. Repeatedly filling the whole buffer with 0 is a perf issue and the first thing to be optimized. 

[Exception](https://isocpp.github.io/CppCoreGuidelines/CppCoreGuidelines.html#es23-prefer-the--initializer-syntax:~:text=Exception%20If%20you%20are,be%20treated%20with%20suspicion.)

Exception If you are declaring an object that is just about to be initialized from input, initializing it would cause a double initialization. However, beware that this might leave uninitialized data beyond the input – and that has been a fertile source of errors and security breaches:
constexpr int max = 8 * 1024;
int buf[max];         // OK, but suspicious: uninitialized
f.read(buf, max);
The cost of initializing that array could be significant in some situations. However, such examples do tend to leave uninitialized variables accessible, so they should be treated with suspicion.

Reply:
If you've got a 10M buffer as a local variable you've got a bigger fundamental problem...

General rule of thumb, if it's more than 1-2K consider using the heap, and pragmatically speaking if much bigger definitely go heap (>4K is my usual rule of thumb but YMMV).

OK I'll buy your exception. I wouldn't do it but we're into the grey "dealer's choice" territory so it's your call